Kernel/Arch/i386/APIC.cpp

```diff
@@ -0,0 +1,128 @@
+#include <AK/Assertions.h>
+#include <AK/Types.h>
+#include <Kernel/Arch/i386/CPU.h>
+#include <Kernel/Arch/i386/APIC.h>
+#include <Kernel/IO.h>
+#include <Kernel/VM/MemoryManager.h>
+
+#define IRQ_APIC_SPURIOUS 0x1f
+
+extern "C" void apic_spurious_interrupt_entry();
+
+asm(
+    ".globl apic_spurious_interrupt_entry \n"
+    "apic_spurious_interrupt_entry: \n"
+    "    iret\n");
+
+namespace APIC {
+
+static volatile u8* g_apic_base = nullptr;
+
+static PhysicalAddress get_base()
+{
+    u32 lo, hi;
+    MSR msr(0x1b);
+    msr.get(lo, hi);
+    return PhysicalAddress(lo & 0xfffff000);
+}
+
+static void set_base(const PhysicalAddress& base)
+{
+    u32 hi = 0;
+    u32 lo = base.get() | 0x800;
+    MSR msr(0x1b);
+    msr.set(lo, hi);
+}
+
+static u32 apic_read(u32 off)
+{
+    return *(volatile u32*)(&g_apic_base[off]);
+}
+
+static void apic_write(u32 off, u32 val)
+{
+    *reinterpret_cast<volatile u32*>(&g_apic_base[off]) = val;
+}
+
+#define APIC_LVT_MASKED (1 << 15)
+#define APIC_LVT_TRIGGER_LEVEL (1 << 14)
+#define APIC_LVT(iv, dm) ((iv & 0xff) | ((dm & 0x7) << 8))
+
+asm(
+    ".globl apic_ap_start \n"
+    ".type apic_ap_start, @function \n"
+    "apic_ap_start: \n"
+    ".set begin_apic_ap_start, . \n"
+    "    jmp apic_ap_start\n" // TODO: implement
+    ".set end_apic_ap_start, . \n"
+    "\n"
+    ".globl apic_ap_start_size \n"
+    "apic_ap_start_size: \n"
+    ".word end_apic_ap_start - begin_apic_ap_start \n");
+
+extern "C" void apic_ap_start(void);
+extern "C" u16 apic_ap_start_size;
+bool init()
+{
+    if (!MSR::have())
+        return false;
+
+    CPUID id(1);
+    if ((id.edx() & (1 << 9)) == 0)
```

Comment:
It would be good to clarify what we're checking for here. Either through a comment, or through use of constants.